src/mito2/src/compaction/twcs.rs

```diff
@@ -84,37 +77,90 @@ impl TwcsPicker {
+            if self.append_mode {
+                // Skip filtering deleted rows once append mode is set.
+                filter_deleted = false;
+            }
 
             if found_runs > max_runs {
                 let files_to_compact = reduce_runs(sorted_runs, max_runs);
-                info!("Building compaction output, active window: {:?}, current window: {}, max runs: {}, found runs: {}, output size: {}, remove deletion markers: {}", active_window, *window,max_runs, found_runs, files_to_compact.len(), filter_deleted);
-                for inputs in files_to_compact {
-                    output.push(CompactionOutput {
-                        output_file_id: FileId::random(),
-                        output_level: LEVEL_COMPACTED, // always compact to l1
-                        inputs,
-                        filter_deleted,
-                        output_time_range: None, // we do not enforce output time range in twcs compactions.
-                    });
+                let files_to_compact_len = files_to_compact.len();
+                info!(
+                    "Building compaction output, active window: {:?}, \
+                        current window: {}, \
+                        max runs: {}, \
+                        found runs: {}, \
+                        output size: {}, \
+                        max output size: {:?}, \
+                        remove deletion markers: {}",
+                    active_window,
+                    *window,
+                    max_runs,
+                    found_runs,
+                    files_to_compact_len,
+                    self.max_output_file_size,
+                    filter_deleted
+                );
+
+                let maybe_enforce_max_output_size = if !filter_deleted
+                    && let Some(max_output_file_size) = self.max_output_file_size
+                {
+                    let enforce_max_output_size =
+                        enforce_max_output_size(files_to_compact, max_output_file_size);
+                    if enforce_max_output_size.len() != files_to_compact_len {
+                        info!("Compaction output file size exceeds threshold {}, split compaction inputs to: {:?}", max_output_file_size, enforce_max_output_size);
+                    }
+                    enforce_max_output_size
+                } else {
+                    files_to_compact
+                };
+
+                for inputs in maybe_enforce_max_output_size {
+                    if inputs.len() > 1 {
+                        output.push(CompactionOutput {
+                            output_file_id: FileId::random(),
+                            output_level: LEVEL_COMPACTED, // always compact to l1
+                            inputs,
+                            filter_deleted,
+                            output_time_range: None, // we do not enforce output time range in twcs compactions.
+                        });
+                    }
```

Comment:
The logic is similar to the logic in the `else if files.files.len() > max_files`. Can we reuse the code? We can invoke `reduce_runs()` or `enforce_file_num` in different branches to get `files_to_compact`.
```rust
let files_to_compact = if found_runs > max_runs {
    let files_to_compact = reduce_runs(sorted_runs, max_runs);
    // ...
    files_to_compact
} else if files.files.len() > max_files {
    let to_merge = enforce_file_num(&files.files, max_files);
    // ...
    vec![to_merge]
} else {
    return;
}

let inputs = if !filter_deleted && let Some(max_output_file_size) = self.max_output_file_size {
    enforce_max_output_size()
} else {
    inputs
};
for input in inputs {
    // ...
}
```

Reply:
> The logic is similar to the logic in the `else if files.files.len() > max_files`. Can we reuse the code? We can invoke `reduce_runs()` or `enforce_file_num` in different branches to get `files_to_compact`.
> 
> ```rust
> let files_to_compact = if found_runs > max_runs {
>     let files_to_compact = reduce_runs(sorted_runs, max_runs);
>     // ...
>     files_to_compact
> } else if files.files.len() > max_files {
>     let to_merge = enforce_file_num(&files.files, max_files);
>     // ...
>     vec![to_merge]
> } else {
>     return;
> }
> 
> let inputs = if !filter_deleted && let Some(max_output_file_size) = self.max_output_file_size {
>     enforce_max_output_size()
> } else {
>     inputs
> };
> for input in inputs {
>     // ...
> }
> ```

fixed in 2aa6d4263a